### PR TITLE
fix: resource reloading

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

Completely reload resources on new or update tenant.
Removed test value for password reset & email verification token lifetime

## Related issues

- https://github.com/supertokens/supertokens-core/pull/555

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/main/java/io/supertokens/ResourceDistributor.java [hunk 1]
```diff
@@ -104,12 +104,27 @@ public synchronized SingletonResource setResource(TenantIdentifier tenantIdentif
         return resource;
     }
 
+    public synchronized SingletonResource removeResource(TenantIdentifier tenantIdentifier,
+                                                      @Nonnull String key) {
+        SingletonResource singletonResource = resources.get(new KeyClass(tenantIdentifier, key));
+        if (singletonResource == null) {
+            return null;
+        }
+        resources.remove(new KeyClass(tenantIdentifier, key), singletonResource);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/ResourceDistributor.java:113]: this only requires a key - why giving both?

File: src/main/java/io/supertokens/featureflag/FeatureFlag.java [hunk 2]
```diff
@@ -125,27 +125,13 @@ public static void initForBaseTenant(Main main, String eeFolderPath) throws Malf
     public static void loadForAllTenants(Main main, List<AppIdentifier> apps) {
         try {
             main.getResourceDistributor().withResourceDistributorLock(() -> {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/featureflag/FeatureFlag.java:127]: change all things to reuse unchanged tenants

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java [hunk 3]
```diff
@@ -94,7 +94,7 @@ public void updateEnabledFeaturesValueReadFromDbTime(long newTime) {
     public void constructor(Main main, AppIdentifier appIdentifier) {
         this.main = main;
         this.appIdentifier = appIdentifier;
-        Cronjobs.addCronjob(main, EELicenseCheck.getInstance(main, this.appIdentifier.getAsPublicTenantIdentifier()));
+        Cronjobs.addCronjob(main, EELicenseCheck.getNewInstance(main, this.appIdentifier.getAsPublicTenantIdentifier()));
```
Comment (rishabhpoddar) [inline ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java:97]: Let's make this cronjob global and have it run one per app instead of one cronjob per app

File: src/main/java/io/supertokens/config/Config.java [hunk 4]
```diff
@@ -110,16 +116,32 @@ public static void loadAllTenantConfig(Main main, TenantConfig[] tenants)
         // At this point, we know that all configs are valid.
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/config/Config.java:116]: verify that if the base tenant of an app's config has changed, then we also reload all the sub tenants (if required)

File: src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java [hunk 5]
```diff
@@ -193,7 +195,14 @@ private void refreshCronjobs() {
 
     public TenantConfig[] getAllTenants() {
         try {
-            return main.getResourceDistributor().withResourceDistributorLockWithReturn(() -> this.tenantConfigs);
+            return main.getResourceDistributor().withResourceDistributorLockWithReturn(() -> {
+                TenantConfig[] tenantConfigs = new TenantConfig[this.tenantConfigs.length];
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:199]: add comment as to why this copy is happening

File: src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java [hunk 6]
```diff
@@ -84,80 +85,81 @@ private TenantConfig[] getAllTenantsFromDb() throws StorageQueryException {
...
-                        Map<TenantIdentifier, TenantConfig> fromDb = new HashMap<>();
-                        for (TenantConfig t : tenantsFromDb) {
-                            fromDb.put(t.tenantIdentifier, t);
-                        }
-                        for (TenantConfig t : this.tenantConfigs) {
-                            TenantConfig fromDbConfig = fromDb.get(t.tenantIdentifier);
-                            if (!t.deepEquals(fromDbConfig)) {
-                                hasChanged = true;
-                                break;
-                            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:106]: need to propagate changes to base app tenant to all tenants...

File: src/main/java/io/supertokens/storageLayer/StorageLayer.java [hunk 7]
```diff
@@ -202,7 +208,7 @@ public static void loadAllTenantStorage(Main main, TenantConfig[] tenants)
                 String userPoolId = storage.getUserPoolId();
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storageLayer/StorageLayer.java:208]: revisit this - it's wrong

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java (hunk 5 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:199]: Done

File: src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java (hunk 6 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:106]: Done

File: src/main/java/io/supertokens/config/Config.java (hunk 4 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/config/Config.java:116]: Done

File: ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java (hunk 3 above)
Comment (sattvikc) [reply ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java:97]: Done

File: ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java [hunk 8]
```diff
@@ -94,7 +93,7 @@ public void updateEnabledFeaturesValueReadFromDbTime(long newTime) {
     public void constructor(Main main, AppIdentifier appIdentifier) {
         this.main = main;
         this.appIdentifier = appIdentifier;
-        Cronjobs.addCronjob(main, EELicenseCheck.getNewInstance(main, this.appIdentifier.getAsPublicTenantIdentifier()));
+        EELicenseCheck.getOrCreateInstance(main);
```
Comment (rishabhpoddar) [inline ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java:96]: remove this from here and put it in the Main.java like the other cronjobs

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java [hunk 9]
```diff
@@ -84,80 +85,85 @@ private TenantConfig[] getAllTenantsFromDb() throws StorageQueryException {
...
                     }
 
+                    boolean sameNumberOfTenants = tenantsFromDb.length == this.tenantConfigs.length;
+
                     this.tenantConfigs = tenantsFromDb;
-                    if (!hasChanged) {
-                        return;
+                    if (tenantsThatChanged.size() == 0 && sameNumberOfTenants) {
+                        return tenantsThatChanged;
                     }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:117]: add a process state here and in the test for when we check that there are not unnecessary reloads, check that process state is null

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java (hunk 8 above)
Comment (sattvikc) [reply ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java:96]: This is task class is not accessible from main

File: ee/src/main/java/io/supertokens/ee/cronjobs/EELicenseCheck.java [hunk 10]
```diff
@@ -6,36 +6,36 @@
...
-    }
+    private static EELicenseCheck instance = null;
 
-    public static EELicenseCheck getNewInstance(Main main, TenantIdentifier tenantIdentifier) {
-        try {
-            EELicenseCheck existingCronTask = (EELicenseCheck) main.getResourceDistributor()
-                    .getResource(tenantIdentifier, RESOURCE_KEY);
+    private EELicenseCheck(Main main, List<List<TenantIdentifier>> tenantsInfo) {
+        super("EELicenseCheck", main, tenantsInfo, true);
+    }
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply ee/src/main/java/io/supertokens/ee/cronjobs/EELicenseCheck.java:23]: Done

File: src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java [hunk 11]
```diff
@@ -91,16 +91,20 @@ public List<TenantIdentifier> refreshTenantsInCoreIfRequired(boolean reloadAllRe
                 try {
                     TenantConfig[] tenantsFromDb = getAllTenantsFromDb();
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:93]: Done

[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:93]: yes

File: src/main/java/io/supertokens/storageLayer/StorageLayer.java [hunk 12]
```diff
@@ -244,7 +244,7 @@ public static void loadAllTenantStorage(Main main, TenantConfig[] tenants, List<
                     String userPoolId = currStorage.getUserPoolId();
                     String connectionPoolId = currStorage.getConnectionPoolId();
                     String uniqueId = userPoolId + "~" + connectionPoolId;
-                    if (idToExistingStorageLayerMap.containsKey(uniqueId)) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storageLayer/StorageLayer.java:241]: Done

File: src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java (hunk 9 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java:117]: Done

File: src/main/java/io/supertokens/cronjobs/Cronjobs.java [hunk 13]
```diff
@@ -92,7 +92,9 @@ public static void addCronjob(Main main, CronTask task) {
         synchronized (instance.lock) {
             instance.executor.scheduleWithFixedDelay(task, task.getInitialWaitTimeSeconds(),
                     task.getIntervalTimeSeconds(), TimeUnit.SECONDS);
-            instance.tasks.add(task);
+            if (!instance.tasks.contains(task)) {
+                instance.tasks.add(task);
+            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/cronjobs/Cronjobs.java:97]: Add a test for this

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/cronjobs/Cronjobs.java (hunk 13 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/cronjobs/Cronjobs.java:98]: Done

File: ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java [hunk 14]
```diff
@@ -93,7 +94,9 @@ public void updateEnabledFeaturesValueReadFromDbTime(long newTime) {
     public void constructor(Main main, AppIdentifier appIdentifier) {
         this.main = main;
         this.appIdentifier = appIdentifier;
-        EELicenseCheck.getOrCreateInstance(main);
+
+        Cronjobs.addCronjob(main, EELicenseCheck.init(main, StorageLayer.getTenantsWithUniqueUserPoolId(main)));
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply ee/src/main/java/io/supertokens/ee/EEFeatureFlag.java:100]: Done
